devito/operator.py

```diff
@@ -208,6 +216,31 @@ def objects(self):
 
     # Compilation
 
+    def _add_implicit(self, expressions):
+        """Create and add any associated implicit expressions."""
+        processed = []
+        for e in expressions:
+            if e.subdomain:
+                try:
+                    dims = [d.root for d in e.free_symbols if isinstance(d, Dimension)]
+                    sub_dims = [d.root for d in e.subdomain.dimensions]
+                    dims = [d for d in dims if d not in frozenset(sub_dims)]
+                    dims.append(e.subdomain.implicit_dimension)
+                    if e._subdomain._implicit_exprs is None:
+                        e._subdomain._implicit_exprs = \
```

Comment:
One of the things that must be fixed and that I hadn't realised before is that here we are changing an Eq's state, and even worse, we're doing it explicitly from outside an Eq. But Eqs are immutable objects, so this should never happen, otherwise there may be unexpected consequences (eg if used as dictionary keys). 

I'll take care of this next week